Bumping up to work with ES 2.3.3, SG-2 and SG-SSL

ewolinetz opened the pull request:
@jcantrill PTAL

jcantrill commented on src/main/java/io/fabric8/elasticsearch/plugin/ConfigurationSettings.java:
```diff
@@ -22,9 +22,6 @@
 	static final String SEARCHGUARD_AUTHENTICATION_PROXY_HEADER = "searchguard.authentication.proxy.header";
 	static final String SEARCHGUARD_CONFIG_INDEX_NAME = "searchguard.config_index_name";
 
-	static final String SEARCHGUARD_TYPE = "ac";
```
With the removal of these is there an 'upgrade' scenario we need to be concerned about?

ewolinetz replied:
we shouldn't need to, we rebuild the SG ac for a node whenever we start it up since the SG config is tied to the pod name

jcantrill commented on src/main/java/io/fabric8/elasticsearch/plugin/acl/DynamicACLFilter.java:
```diff
@@ -65,56 +80,78 @@
-	private static final String SEARCHGUARD_TYPE = "ac";
-	private static final String SEARCHGUARD_ID = "ac";
-	private static final String[] BLACKLIST_PROJECTS = {"default", "openshift", "openshift-infra" };
+	private static final String[] BLACKLIST_PROJECTS = {"default", "openshift", "openshift-infra", "kube-system" };
```
Should this be configurable?  I dont recall if these are just the defaults, but we may have need to add to this list.

ewolinetz replied:
it should be.. i mistakenly thought we had that in master; i'll update to add that in i believe i have it stashed in another branch

jcantrill commented on src/main/java/io/fabric8/elasticsearch/plugin/acl/DynamicACLFilter.java:
```diff
@@ -65,56 +80,78 @@
+                .builder()
+                .put("path.home", ".")
+                .put("path.conf", ".")
+                .put(SSLConfigConstants.SEARCHGUARD_SSL_TRANSPORT_KEYSTORE_FILEPATH, "/usr/share/elasticsearch/config/admin.jks")
```
API here is load from settings or use the given default?

ewolinetz replied:
this can be moved into a setting

jcantrill commented on src/main/java/io/fabric8/elasticsearch/plugin/acl/DynamicACLFilter.java:
```diff
@@ -238,98 +274,135 @@ private synchronized void syncAcl() {
-		request.putInContext(OS_ES_REQ_ID, ACL_FILTER_ID);
-		GetResponse response = esClient.get(request).actionGet(); // need to worry about timeout?
-		return mapper.readValue(response.getSourceAsBytes(), SearchGuardACL.class);
+		GetResponse response = esClient.get(getRequest).actionGet(); // need to worry about timeout?
```
We probably will only be able to answer your question based on observation.  The previous impl utilized the lock and notifier because of the 'stop the world' threading on incoming requests.  I think the approach we took was to make a best effort to get what we wanted and then defer to what was cached and/or most restrictive.

jcantrill commented on src/main/java/io/fabric8/elasticsearch/plugin/acl/DynamicACLFilter.java:
```diff
@@ -238,98 +274,135 @@ private synchronized void syncAcl() {
 	
+	private SearchGuardRolesMapping readRolesMappingACL(Client esClient) throws IOException {
+		
+		GetRequest getRequest = esClient.prepareGet("searchguard", "rolesmapping", "0")
```
Do we need to be concerned about these values changing based on configuration?  Should they be moved to constants?

ewolinetz replied:
I'll update them to be constants.  I can see them changing only if Searchguard changed where they look for them; so we should be safe for a while

jcantrill commented on src/main/java/io/fabric8/elasticsearch/plugin/acl/DynamicACLFilter.java:
```diff
@@ -238,98 +274,135 @@ private synchronized void syncAcl() {
-		}
+	private void writeACL(Client esClient, SearchGuardRoles roles, SearchGuardRolesMapping rolesMapping) throws IOException {
+		
+		IndexRequest rolesIR = new IndexRequest("searchguard").type("roles").id("0").refresh(true)
```
Configuration and constants?

jcantrill commented on src/main/java/io/fabric8/elasticsearch/plugin/acl/DynamicACLFilter.java:
```diff
@@ -238,98 +274,135 @@ private synchronized void syncAcl() {
+		if ( cur.getNodes().length > 0 )
+			logger.debug("Successfully reloaded config with '{}' nodes", cur.getNodes().length);
+		else
+			logger.warn("Failed to reloaded config -- '{}' nodes updated", cur.getNodes().length);
```
Do we need string substitution here?  Isn't the only option 0?

ewolinetz replied:
it was more of a sanity check; I can update it

jcantrill commented on src/main/java/io/fabric8/elasticsearch/plugin/acl/DynamicACLFilter.java:
```diff
@@ -238,98 +274,135 @@ private synchronized void syncAcl() {
+			
+			String cd = settings.get(SG_CONFIG_SETTING_PATH, DEFAULT_SG_CONFIG_SETTING_PATH);
+			// create initial sg config
+			boolean success = uploadFile(esClient, cd+"sg_config.yml", "config");
```
Make configurable or move to constants.  The former is preferable.

jcantrill replied:
There seems to be a loop here that could be more elegant.  Additionally, it seems there is possibility that we have partial success.  What are the ramifications of not uploading all the files?  Also, if we were able to either loop or return something other then boolean, we could report a more specific debug message about which specific upload failed.

ewolinetz replied:
I'll think about how we can do that.  I ripped this from SG[1] since that is how they seed their configuration initially.  I'll try to make it more elegant

[1] https://github.com/floragunncom/search-guard/blob/master/src/main/java/com/floragunn/searchguard/tools/SearchGuardAdmin.java

jcantrill commented on src/main/java/io/fabric8/elasticsearch/plugin/acl/DynamicACLFilter.java:
```diff
@@ -238,98 +274,135 @@ private synchronized void syncAcl() {
 		try {
-			//This should return nothing initially - if it does, we're done
-			acl = loadAcl(esClient);
+			if ( esClient.admin().indices().exists(new IndicesExistsRequest().indices(new String[]{"searchguard"})).actionGet().isExists() ) {
```
configurable or const?

ewolinetz replied:
I forgot about this, we should be pulling it from the configuration.  Currently SG doesn't let us specify the index it looks for [1]

[1] https://github.com/floragunncom/search-guard/issues/176

jcantrill commented on src/main/java/io/fabric8/elasticsearch/plugin/acl/DynamicACLFilter.java:
```diff
@@ -345,4 +418,38 @@ public int order() {
+                            .consistencyLevel(WriteConsistencyLevel.DEFAULT).source(readXContent(reader, XContentType.YAML)))
+                            .actionGet().getId();
+
+            if ("0".equals(id)) {
```
refactor to constant or setting to make it obvious what it is

jcantrill commented on src/main/java/io/fabric8/elasticsearch/plugin/acl/DynamicACLFilter.java:
```diff
@@ -345,4 +418,38 @@ public int order() {
+        return false;
+    }
+
+    private BytesReference readXContent(final Reader reader, final XContentType xContentType) throws IOException {
```
Apache commons has IOUtils that may be able to replace this method; its worth investigating

ewolinetz replied:
this was ripped from SG since it worked... I'll check out IOUtils though

jcantrill commented on src/main/java/io/fabric8/elasticsearch/plugin/acl/RoleBuilder.java:
```diff
@@ -0,0 +1,109 @@
+/**
+ * Copyright (C) 2015 Red Hat, Inc.
```
Update to 2016

jcantrill commented on src/main/java/io/fabric8/elasticsearch/plugin/acl/RoleBuilder.java:
```diff
@@ -0,0 +1,109 @@
+		return this;
+	}
+	
+	public RoleBuilder setCluster(List<String> cluster) {
```
setClusters or clusters.... plural

jcantrill commented on src/main/java/io/fabric8/elasticsearch/plugin/acl/RoleBuilder.java:
```diff
@@ -0,0 +1,109 @@
+	}
+	
+	public RoleBuilder setCluster(String[] clusters) {
+		List<String> clusterList = new ArrayList<String>();
```
Arrays.asList

jcantrill commented on src/main/java/io/fabric8/elasticsearch/plugin/acl/RoleBuilder.java:
```diff
@@ -0,0 +1,109 @@
+			index.setIndex(indexKey);
+			
+			List<Type> types = new ArrayList<Type>();				
+			for ( String typeKey : indices.get(indexKey).keySet() ) {
```
concerns of a nullpointer here?

ewolinetz replied:
it _should_ be alright since we get indexKey from iterating over the keyset...

ewolinetz replied:
Sorry, this didn't really answer your question... it should be alright since whenever we add an indices entry, we put with a new object.  at worst `.keySet()` would return an empty set.

jcantrill commented on src/main/java/io/fabric8/elasticsearch/plugin/acl/RolesMappingBuilder.java:
```diff
@@ -0,0 +1,66 @@
+/**
+ * Copyright (C) 2015 Red Hat, Inc.
```
2016

jcantrill commented on src/main/java/io/fabric8/elasticsearch/plugin/acl/RolesMappingBuilder.java:
```diff
@@ -0,0 +1,66 @@
+
+public class RolesMappingBuilder {
+	
+	private Map<String, HashSet<String>> roles = new HashMap<String, HashSet<String>>();
```
Can simplify to new HashMap<>()

jcantrill commented on src/main/java/io/fabric8/elasticsearch/plugin/acl/RolesMappingBuilder.java:
```diff
@@ -0,0 +1,66 @@
+	private List<RolesMapping> rolesMappings;
+	
+	public RolesMappingBuilder() {
+		rolesMappings = new ArrayList<RolesMapping>();
```
Might as well move this to the field and remove the noarg constructor

jcantrill commented on src/main/java/io/fabric8/elasticsearch/plugin/acl/SearchGuardRoles.java:
```diff
@@ -0,0 +1,255 @@
+/**
+ * Copyright (C) 2015 Red Hat, Inc.
```
2016

jcantrill commented on src/main/java/io/fabric8/elasticsearch/plugin/acl/SearchGuardRoles.java:
```diff
@@ -0,0 +1,255 @@
+		RolesBuilder builder = new RolesBuilder();
+		
+		for ( String key : source.keySet() ) {
+			RoleBuilder roleBuilder = new RoleBuilder().setName(key);
```
Any reason not to make this a constructor arg for brevity?

ewolinetz replied:
not really, had it like this to make it obvious what was being done

jcantrill commented on src/main/java/io/fabric8/elasticsearch/plugin/acl/SearchGuardRoles.java:
```diff
@@ -0,0 +1,255 @@
+			List<String> cluster = (ArrayList<String>)role.get(CLUSTER_HEADER);
+			roleBuilder.setCluster(cluster);
+			
+			HashMap<String, HashMap<String, ArrayList<String>>> indices = (HashMap<String, HashMap<String, ArrayList<String>>>) role.get(INDICES_HEADER);
```
Any change this can be modified while processing thus throwing a modification exception?

ewolinetz replied:
i dont think so unless source can be changed while we're in this method?

jcantrill commented on src/main/java/io/fabric8/elasticsearch/plugin/acl/SearchGuardRolesMapping.java:
```diff
@@ -0,0 +1,152 @@
+			this.users = users;
+		}
+		
+		public void addUser(String user) {
```
Not needed?

ewolinetz replied:
This is used, setUsers might not be needed though

ewolinetz replied:
scratch that, i was definitely thinking about another section of code that used 'addUser'... will remove (also didn't see with your comment that it was a blank method)

jcantrill commented on src/main/java/io/fabric8/elasticsearch/plugin/kibana/KibanaSeed.java:
```diff
@@ -94,6 +109,7 @@ public static void setDashboards(String user, Set<String> projects, Set<String>
 		// If none have been set yet
 		if ( indexPatterns.isEmpty() ) {
 			create(user, sortedProjects, true, esClient, kibanaIndex, kibanaVersion);
+			changed = true;
```
Does this need to be an AtomicBoolean?

jcantrill commented on src/main/java/io/fabric8/elasticsearch/plugin/kibana/KibanaSeed.java:
```diff
@@ -121,8 +143,85 @@ public static void setDashboards(String user, Set<String> projects, Set<String>
+			refreshKibanaUser(user, kibanaIndex, esClient);
+	}
+	
+	private static void refreshKibanaUser(String username, String kibanaIndex, Client esClient) {
```
do we need to synchronize this or impl state and atomics for multi-threaded access?

ewolinetz replied:
i dont believe we'll need to

jcantrill commented on src/main/java/io/fabric8/elasticsearch/plugin/kibana/KibanaSeed.java:
```diff
@@ -121,8 +143,85 @@ public static void setDashboards(String user, Set<String> projects, Set<String>
+		logger.debug("Refreshed '{}' successfully on {} of {} shards", userIndex, response.getSuccessfulShards(), response.getTotalShards());
+	}
+	
+	private static void initialKibanaSearchMapping(String username, String kibanaIndex, Client esClient) {
```
smae here

jcantrill commented on src/main/java/io/fabric8/elasticsearch/plugin/kibana/KibanaSeed.java:
```diff
@@ -121,8 +143,85 @@ public static void setDashboards(String user, Set<String> projects, Set<String>
+			if ( !getResponse.getMappings().get(userIndex).containsKey("search") ) {
+			
+				PutMappingRequest putRequest = new PutMappingRequest().indices(new String[]{userIndex}).type("source").source(
+				"{\"properties\":{\"title\":{\"type\":\"string\"},\"description\":{\"type\":\"string\"},\"hits\":{\"type\":\"integer\"},\"columns\":{\"type\":\"string\"},\"sort\":{\"type\":\"string\"},\"version\":{\"type\":\"integer\"},\"kibanaSavedObjectMeta\":{\"properties\":{\"searchSourceJSON\":{\"type\":\"string\"}}}}}"		
```
I wonder if we should move this to a file for easier updating and editing.  We could statically load it once and then it would mitigate having to do all the escaping

jcantrill commented:
Mostly nits and questions for things to think about around multi-threading scenarios.  Also, add some tests please.

ewolinetz commented:
@jimmidyson the license-check is expecting the year to be "2015" and is failing with me having "2016" should that be updated?

jimmidyson commented:
Up to you. We tend to leave it the same across the project as AFAIK the copyright runs for more than long enough to cover any reasonable time period.

ewolinetz commented:
@jcantrill thoughts on above?

jcantrill commented:
They were new files added this year so the license check should account for
that...but it's really not that important to me

On Thursday, August 4, 2016, Eric Wolinetz <notifications@github.com> wrote:

> @jcantrill <https://github.com/jcantrill> thoughts on above?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/fabric8io/openshift-elasticsearch-plugin/pull/29#issuecomment-237608279>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEVnOLiBeoSb3gnBO0bcWypakAcv251uks5qchQCgaJpZM4JYf7s>
> .
>


-- 
--
Jeff Cantrill
Senior Software Engineer, Red Hat Engineering
OpenShift Integration Services
Red Hat, Inc.
*Office*: 703-748-4420 | 866-546-8970 ext. 8162420
jcantril@redhat.com
http://www.redhat.com

ewolinetz commented:
@jimmidyson lets leave it as is for now then I suppose.
@jcantrill Addressed nits -- for locking I updated to match what we were doing with our previous version.

ewolinetz commented:
added tests, probably need to update readme too... but that may not be necessary in this PR

jcantrill commented:
LGTM...best I can tell

ewolinetz commented:
Okay, we may need to refactor since that other pr was merged... If not and
you're good with this we can merge it.

On Aug 6, 2016 3:28 PM, "Jeff Cantrill" <notifications@github.com> wrote:

LGTM...best I can tell

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/fabric8io/openshift-elasticsearch-plugin/pull/29#issuecomment-238047175>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/ABXIEeayAlyYgsI438et2TvAu-Q_3Hbwks5qdO5agaJpZM4JYf7s>
.

ewolinetz commented:
@jcantrill https://github.com/fabric8io/openshift-elasticsearch-plugin/issues/30 is to not lose the README changes for this PR

ewolinetz commented:
@jcantrill hold off on merging this actually... i just came across a NPE in kibana_seed when deleting a namespace

ewolinetz commented:
@jcantrill PTAL

jcantrill commented on src/main/java/io/fabric8/elasticsearch/plugin/kibana/KibanaSeed.java:
```diff
@@ -407,10 +387,21 @@ private static void executeDelete(String index, String type, String id, Client e
-		} catch (InterruptedException | ExecutionException e) {
+			ActionFuture<DeleteResponse> response = esClient.delete(request);
+			// wait until response isDone()
+			while ( !response.isDone() ) {
```
I think http://docs.oracle.com/javase/8/docs/api/java/util/concurrent/Future.html#get-long-java.util.concurrent.TimeUnit- is what you desire instead of Thread.sleep

ewolinetz replied:
I may be missing a plugin on my client... i'm seeing if fixing the client will let me revert to have `.get()` block instead of receiving a NPE

ewolinetz replied:
Updating the try-catch to be
```
try {
  DeleteResponse response = esClient.prepareDelete(...).get();
catch { 
...
}
```

gets us this stack-trace:
```
java.lang.NullPointerException
	at org.elasticsearch.action.ActionWriteResponse.writeTo(ActionWriteResponse.java:57)
	at org.elasticsearch.action.delete.DeleteResponse.writeTo(DeleteResponse.java:101)
	at org.elasticsearch.transport.netty.NettyTransportChannel.sendResponse(NettyTransportChannel.java:97)
	at org.elasticsearch.transport.netty.NettyTransportChannel.sendResponse(NettyTransportChannel.java:75)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$OperationTransportHandler$1.onResponse(TransportReplicationAction.java:248)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$OperationTransportHandler$1.onResponse(TransportReplicationAction.java:244)
	at org.elasticsearch.action.support.TransportAction$ResponseFilterChain.proceed(TransportAction.java:212)
	at com.floragunn.searchguard.filter.SearchGuardFilter.apply(SearchGuardFilter.java:171)
	at org.elasticsearch.action.support.TransportAction$ResponseFilterChain.proceed(TransportAction.java:210)
	at io.fabric8.elasticsearch.plugin.KibanaUserReindexAction.apply(KibanaUserReindexAction.java:162)
	at org.elasticsearch.action.support.TransportAction$ResponseFilterChain.proceed(TransportAction.java:210)
	at org.elasticsearch.action.support.TransportAction$FilteredActionListener.onResponse(TransportAction.java:237)
	at org.elasticsearch.action.support.TransportAction$FilteredActionListener.onResponse(TransportAction.java:223)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$ReroutePhase.finishOnSuccess(TransportReplicationAction.java:589)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$ReroutePhase$1.handleResponse(TransportReplicationAction.java:514)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$ReroutePhase$1.handleResponse(TransportReplicationAction.java:500)
	at org.elasticsearch.transport.TransportService$DirectResponseChannel.processResponse(TransportService.java:819)
	at org.elasticsearch.transport.TransportService$DirectResponseChannel.sendResponse(TransportService.java:803)
	at org.elasticsearch.transport.TransportService$DirectResponseChannel.sendResponse(TransportService.java:793)
	at org.elasticsearch.transport.DelegatingTransportChannel.sendResponse(DelegatingTransportChannel.java:58)
	at org.elasticsearch.transport.RequestHandlerRegistry$TransportChannelWrapper.sendResponse(RequestHandlerRegistry.java:134)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$ReplicationPhase.doFinish(TransportReplicationAction.java:988)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$ReplicationPhase.doRun(TransportReplicationAction.java:860)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$PrimaryPhase.finishAndMoveToReplication(TransportReplicationAction.java:715)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$PrimaryPhase.doRun(TransportReplicationAction.java:659)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$PrimaryOperationTransportHandler.messageReceived(TransportReplicationAction.java:279)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$PrimaryOperationTransportHandler.messageReceived(TransportReplicationAction.java:271)
	at com.floragunn.searchguard.ssl.transport.SearchGuardSSLTransportService.messageReceivedDecorate(SearchGuardSSLTransportService.java:161)
	at com.floragunn.searchguard.transport.SearchGuardTransportService.messageReceivedDecorate(SearchGuardTransportService.java:190)
	at com.floragunn.searchguard.ssl.transport.SearchGuardSSLTransportService$Interceptor.messageReceived(SearchGuardSSLTransportService.java:100)
	at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:75)
	at org.elasticsearch.transport.TransportService$4.doRun(TransportService.java:376)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

I still see that the record was deleted... it just couldn't get me a proper reponse

ewolinetz commented:
@jcantrill found the source of the issue for the NPE. PTAL

jcantrill commented:
LGTM after squash

ewolinetz commented:
squashed

ewolinetz commented:
[merge]
